blake3-air/src/constants.rs

```diff
@@ -0,0 +1,27 @@
+pub const BITS_PER_LIMB: usize = 16;
+pub const U32_LIMBS: usize = 32 / BITS_PER_LIMB;
+
+// The constants from the reference implementation.
+// Saved as pairs of 16 bit integers in [lo, hi] format.
+pub(crate) const IV: [[u32; 2]; 8] = [
+    [0xE667, 0x6A09],
+    [0xAE85, 0xBB67],
+    [0xF372, 0x3C6E],
+    [0xF53A, 0xA54F],
+    [0x527F, 0x510E],
+    [0x688C, 0x9B05],
+    [0xD9AB, 0x1F83],
+    [0xCD19, 0x5BE0],
+];
+
+/// The index map for the permutation used to permute the block words.
+const MSG_PERMUTATION: [usize; 16] = [2, 6, 3, 10, 7, 0, 4, 13, 1, 11, 12, 5, 9, 14, 15, 8];
+
+/// Apply the MSG_PERMUTATION to an array.
+pub(crate) fn permute<T: Clone>(m: &mut [T; 16]) {
+    let mut permuted = m.clone();
```

Comment:
Do you think performance is important here? If so we could use cycle decomposition to apply the permutation in-place. But at least it's a fixed-length array so no allocation.

Reply:
I guess here there are two 8-cycles, `0 2 3 10 12 9 11 5` and `1 6 4 7 13 14 15 8`, so we'd do two "rotations". Maybe the compiled code could end up the same anyway though...

Reply:
Hmm, I had a go at doing this and couldn't see a difference. It's also definitely not a bottleneck, even if I replace it by a function which does nothing the proof speed is essentially unchanged.

I'll leave it as it currently is but add a note in the code just in case this is something we want to come back to in future for some reason.